osu.Game/Online/Chat/MessageNotifier.cs

```diff
@@ -0,0 +1,257 @@
+                        notificationOverlay?.Post(notification);
+                    }
+                }
+            }
+        }
+
+        private static string[] getWords(string input) => input.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
+
+        /// <summary>
+        /// Finds the first matching string/word in both <paramref name="x"/> and <paramref name="y"/> (case-insensitive)
+        /// </summary>
+        private static string hasCaseInsensitive(IEnumerable<string> x, IEnumerable<string> y) => x.FirstOrDefault(x2 => anyCaseInsensitive(y, x2));
+
+        private static bool anyCaseInsensitive(IEnumerable<string> x, string y) => x.Any(x2 => x2.Equals(y, StringComparison.OrdinalIgnoreCase));
+
+        public class HighlightNotification : SimpleNotification
+        {
+            public HighlightNotification(string highlighter, string word, Action onClick)
+            {
+                Icon = FontAwesome.Solid.Highlighter;
+                Text = $"'{word}' was mentioned in chat by '{highlighter}'. Click to find out why!";
+                this.onClick = onClick;
+            }
+
+            private readonly Action onClick;
+
+            public override bool IsImportant => false;
+
+            [BackgroundDependencyLoader]
+            private void load(OsuColour colours)
+            {
+                IconBackgound.Colour = colours.PurpleDark;
+                Activated = delegate
+                {
+                    onClick?.Invoke();
+                    return true;
+                };
+            }
+        }
+
+        public class PrivateMessageNotification : SimpleNotification
+        {
+            public PrivateMessageNotification(string username, Action onClick)
+            {
+                Icon = FontAwesome.Solid.Envelope;
+                Username = username;
+                MessageCount = 1;
+                this.onClick = onClick;
+            }
+
+            private int messageCount;
+
+            public int MessageCount
+            {
+                get => messageCount;
+                set => Text = (messageCount = value) > 1 ? $"You received {messageCount} private messages from '{Username}'. Click to read it!" : $"You received a private message from '{Username}'. Click to read it!";
+            }
+
+            public string Username { get; set; }
+
+            private readonly Action onClick;
+
+            public override bool IsImportant => false;
+
+            [BackgroundDependencyLoader]
+            private void load(OsuColour colours, MessageNotifier notifier)
+            {
+                IconBackgound.Colour = colours.PurpleDark;
+                Activated = delegate
+                {
+                    onClick?.Invoke();
+
+                    if (notifier.privateMessageNotifications.Contains(this))
+                        notifier.privateMessageNotifications.Remove(this);
```

Comment:
I tried to make NotificationOverlay expose its notifications  so I can manipulate them later, that didn't work and a list containing all instances of the notification was working. To be sure that we don't change notifications that aren't even inside the NotificationOverlay anymore I made it to remove itself. I would be glad if you could guide on how to change this.

Reply:
Shouldn't you just include these 2 lines inside the `onClick` action.

Reply:
> Shouldn't you just include these 2 lines inside the `onClick` action.

Already did that with the suggestion of un-god-ing the HandleMessages method.

Reply:
@bdach any thoughts?

Reply:
I suppose if it's in one central place instead of present in one and absent in two others I can temporarily live with it, but it'd be better to have it gone.